fix(nuxt): resolve internal `target: blank` links with base

Jannchie opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxt.com/docs/community/contribution
-->

### 🔗 Linked issue

<!-- Please ensure there is an open issue and mention its number as #123 -->

- resolve #23737 

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

Now, such a link will automatically append the baseURL.
``` vue
  <div>
    <NuxtLink to="/about" target="_blank">link with target</NuxtLink>
  </div>
```

This should be a breaking change. because it's possible that someone could explicitly add baseURL to the `to` attribute to get around the bug.

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

Jannchie committed:
fix(link): `target=_blank` nuxt link should join baseURL

Jannchie committed:
fix(nuxt): import useRuntimeConfig

Jannchie committed:
Merge branch 'main' into fix/nuxt-link

Jannchie committed:
fix(link): type check

Jannchie committed:
Merge branch 'fix/nuxt-link' of github.com:Jannchie/nuxt into fix/nuxt-link

Mini-ghost commented:
Hi! I happen to be researching this issue as well. Would you like to consider some other ideas?

```ts
let href = typeof to.value === 'object' ? router.resolve(to.value)?.href ?? null : to.value || null

if (href && href.startsWith('/') && !href.startsWith('//')) {
  const _base = withLeadingSlash(withTrailingSlash(baseURL))
  if (_base !== '/' && !href starts with(_base)) {
    href = joinURL(_base, href)
  }
}
```

This way, the following examples will work as expected:

```vue
<template>
  <NuxtLink to="http://localhost:3000/app/about" target="_blank">
    link with target and protocol
  </NuxtLink>
  <!--
    Output:
    <a href="http://localhost:3000/app/about" rel="noopener noreferrer" target="_blank"> link with target and baseURL </a>
  -->

  <NuxtLink to="//localhost:3000/app/about" target="_blank">
    link with target and starts with `//`
  </NuxtLink>
  <!--
    Output:
    <a href="//localhost:3000/app/about" rel="noopener noreferrer" target="_blank"> link with target and baseURL </a>
  -->

  <NuxtLink to="/app/about" target="_blank">
    link with target and baseURL
  </NuxtLink>
  <!--
    Output:
    <a href="/app/about" rel="noopener noreferrer" target="_blank"> link with target and baseURL </a>
  -->

  <NuxtLink to="/about" target="_blank">
    link with target without baseURL
  </NuxtLink>
  <!--
    Output:
    <a href="/app/about" rel="noopener noreferrer" target="_blank"> link without target </a>
  -->
</template>

```

Jannchie committed:
fix(link): pass unit test

Jannchie committed:
fix(link): ignore href start with baseURL

Jannchie commented:
> Hi! I happen to be researching this issue as well. Would you like to consider some other ideas?

Yeah, your idea is better. I modified this pull request with reference to your implementation... Thank you!

Mini-ghost commented on packages/nuxt/src/app/components/nuxt-link.ts:
```diff
@@ -280,8 +279,15 @@ export function defineNuxtLink (options: NuxtLinkOptions) {
+        let href = typeof to.value === 'object' ? router.resolve(to.value)?.href ?? null : to.value || null
 
+        // joins with `baseURL` if it's an relative link
+        if (import.meta.client && href && href.startsWith('/') && !href.startsWith('//')) {
```
Why check only on the client side? This could lead to inconsistent results between the server side and the client side.

```suggestion
-        if (import.meta.client && href && href.startsWith('/') && !href.startsWith('//')) {
+        if (href && href.startsWith('/') && !href.startsWith('//')) {
```

Mini-ghost replied:
Could you please review it once more? I haven't personally encountered the situation where `useRuntimeConfig` is unable to retrieve the `baseURL` on the server side in my environment.

Jannchie replied:
Have you tried this command?

```
pnpm vitest nuxt-link
```

In my environment, the test does not pass without detecting `import.meta.client`:

```
 FAIL  packages/nuxt/test/nuxt-link.test.ts > nuxt-link:propsOrAttributes > `isExternal` is `true` > href > resolves route location object
Error: [nuxt] instance unavailable
 ❯ useNuxtApp packages/nuxt/src/app/nuxt.ts:395:13
    409| }
    410|
    411| export function defineAppConfig<C extends AppConfigInput> (config: C): C {
       |                                             ^
    412|   return config
    413| }
 ❯ Module.useRuntimeConfig packages/nuxt/src/app/nuxt.ts:404:10
 ❯ packages/nuxt/src/app/components/nuxt-link.ts:286:27
 ❯ nuxtLink packages/nuxt/test/nuxt-link.test.ts:46:96
 ❯ packages/nuxt/test/nuxt-link.test.ts:115:16

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[3/3]⎯

 Test Files  1 failed (1)
      Tests  3 failed | 25 passed (28)
   Start at  14:31:52
   Duration  2.81s (transform 179ms, setup 2.22s, collect 167ms, tests 13ms, environment 0ms, prepare 155ms)
```

danielroe reviewed (changes requested):
Would you be able to add some tests to cover this?

Jannchie committed:
Update packages/nuxt/src/app/components/nuxt-link.ts

Co-authored-by: Alex Liu <dsa1314@gmail.com>

Jannchie commented on packages/nuxt/src/app/components/nuxt-link.ts:
```diff
@@ -280,8 +279,15 @@ export function defineNuxtLink (options: NuxtLinkOptions) {
+        let href = typeof to.value === 'object' ? router.resolve(to.value)?.href ?? null : to.value || null
 
+        // joins with `baseURL` if it's an relative link
+        if (import.meta.client && href && href.startsWith('/') && !href.startsWith('//')) {
```
Have you tried to test it by `pnpm vitest nuxt-link`?

In test cases, it will report an error for lack of context.

May be we should fix it by:

```suggestion
        if (hasInjectionContext() && href && href.startsWith('/') && !href.startsWith('//')) {
```

Mini-ghost replied:
Have you ever attempted to include this in tests?

```
// nuxt-link.test.ts
// Mocks Nuxt `useRuntimeConfig()`
vi.mock('../src/app/nuxt', () => ({
  useRuntimeConfig: () => ({
    app: {
      baseURL: '/'
    }
  })
}))
```

Jannchie replied:
I see, thank you for pointing out the correct approach!

Jannchie committed:
Merge branch 'main' into fix/nuxt-link

Jannchie committed:
chore(test): add text case & mock

Jannchie committed:
Merge branch 'fix/nuxt-link' of github.com:Jannchie/nuxt into fix/nuxt-link

Jannchie commented:
> Would you be able to add some tests to cover this?

I've added the test cases!

lihbr commented on packages/nuxt/src/app/components/nuxt-link.ts:
```diff
@@ -280,8 +279,15 @@ export function defineNuxtLink (options: NuxtLinkOptions) {
+        // joins with `baseURL` if it's an relative link
+        if (href && href.startsWith('/') && !href.startsWith('//')) {
+          const baseURL = useRuntimeConfig().app.baseURL
+          if (href !== '/' && !href.startsWith(baseURL)) {
```
What happens if it's a blank target to the homepage there? 🤔

Or I assume we want to check `base` here instead?

Jannchie replied:
The '/' doesn't automatically append the baseURL with the current practice, which is indeed a problem.

But then I realized, maybe someone really needs to navigate to a path without baseURL?
Current practice does not seem to allow for this except by using full paths.

lihbr replied:
OK, I think that makes sense, then we just need to update the following for now to my understanding:
```suggestion
          if (baseURL !== '/' && !href.startsWith(baseURL)) {
```

lihbr commented on packages/nuxt/test/nuxt-link.test.ts:
```diff
@@ -125,6 +135,21 @@ describe('nuxt-link:propsOrAttributes', () => {
         expect(nuxtLink({ to: 'https://nuxtjs.org' }).props.target).toBe(null)
       })
+
+      it('target="_blank" with baseURL', () => {
```
Small style suggestion:
```
prefixes target="_blank" internal links with baseURL
```

lihbr commented on packages/nuxt/test/nuxt-link.test.ts:
```diff
@@ -125,6 +135,21 @@ describe('nuxt-link:propsOrAttributes', () => {
+        }, () => {
+          expect(nuxtLink({ to: '/to', target: '_blank' }).props.href).toBe('/base/to')
+          expect(nuxtLink({ to: '/base/to', target: '_blank' }).props.href).toBe('/base/to')
+          expect(nuxtLink({ to: 'http://nuxtjs.org/app/about', target: '_blank' }).props.href).toBe('http://nuxtjs.org/app/about')
+          expect(nuxtLink({ to: '//nuxtjs.org/app/about', target: '_blank' }).props.href).toBe('//nuxtjs.org/app/about')
```
Maybe it'd be semantically more correct to check for external links NOT to include the baseURL here.

lihbr reviewed (changes requested):
LGTM! just a small concern with the checks, and some optional style suggestions for the tests :)

Thanks for making this!

Jannchie committed:
Update packages/nuxt/src/app/components/nuxt-link.ts

Co-authored-by: Lucie <25330882+lihbr@users.noreply.github.com>

Jannchie committed:
chore(test): rename & separate test case

lihbr reviewed (approved):
LGTM, thanks!

lihbr committed:
chore: typo

danielroe committed:
Merge remote-tracking branch 'origin/main' into fix/nuxt-link

danielroe committed:
fix: resolve urls consistently with and without `_target: 'blank'`

- move `useRuntimeConfig()` out of render function
- respect `trailingSlashBehavior`
- don't reimplement `hasProtocol`

danielroe committed:
fix: assert type
